File: backend/app/schemas/entitlements.py
"""ARCH-30 Tranche 2 (D-8) — add-on entitlement responses."""

from __future__ import annotations

import uuid
from datetime import datetime
from typing import Literal, Optional

from pydantic import BaseModel, Field

AddonState = Literal["ACTIVE", "GRACE", "LAPSED", "NOT_GRANTED"]


class AddonAccessResponse(BaseModel):
    addon_key: str
    display_name: str
    description: str
    state: AddonState
    source: Optional[Literal["TIER", "SUBSCRIPTION"]] = None
    grace_ends_at: Optional[datetime] = Field(
        default=None,
        description=(
            "End of full access during a downgrade grace. Provisional until the "
            "sweep stamps it, which happens within one sweep interval."
        ),
    )
    can_create: bool
    can_maintain: bool
    live_resource_count: int
    halt_effect: str
    monthly_price_micros: int
    currency: str
    purchasable: bool = Field(
        description="True when an owner can buy this add-on self-serve right now."
    )
    included_in: list[str] = Field(
        default_factory=list,
        description="Published plans that include this add-on.",
    )


class OrganizationEntitlementsResponse(BaseModel):
    organization_id: uuid.UUID
    as_of: datetime
    addons: list[AddonAccessResponse]


__all__ = ["AddonAccessResponse", "AddonState", "OrganizationEntitlementsResponse"]

File: frontend/src/types/entitlements.ts
/**
 * ARCH-30 Tranche 2 (D-8) — add-on entitlements.
 *
 * Mirrors `app/schemas/entitlements.py`. Every boolean the console acts on —
 * `can_create`, `can_maintain`, `purchasable` — is computed by the backend and
 * sent. The console never re-derives them from `state`, so a button can never
 * be enabled over an endpoint that answers 402.
 */

export type AddonKey = "addon.custom_domain" | "addon.warehouse_sync";

export type AddonState = "ACTIVE" | "GRACE" | "LAPSED" | "NOT_GRANTED";

export interface AddonAccess {
  readonly addon_key: AddonKey;
  readonly display_name: string;
  readonly description: string;
  readonly state: AddonState;
  readonly source: "TIER" | "SUBSCRIPTION" | null;
  readonly grace_ends_at: string | null;
  readonly can_create: boolean;
  readonly can_maintain: boolean;
  readonly live_resource_count: number;
  readonly halt_effect: string;
  readonly monthly_price_micros: number;
  readonly currency: string;
  readonly purchasable: boolean;
  readonly included_in: readonly string[];
}

export interface OrganizationEntitlements {
  readonly organization_id: string;
  readonly as_of: string;
  readonly addons: readonly AddonAccess[];
}

/** The structured 402 body from `app/api/addon_gate.py`. */
export interface AddonRequiredDetail {
  readonly code: "ADDON_REQUIRED";
  readonly addon_key: AddonKey;
  readonly addon_name: string;
  readonly state: AddonState;
  readonly grace_ends_at: string | null;
  readonly message: string;
}
